docs/authentication.md

```diff
@@ -0,0 +1,318 @@
+
+## Patterns
+
+The first step to identifying which authentication pattern you need is understanding the data-fetching strategy you want. We can then determine which authentication providers support this strategy. There are two main strategies:
+
+- Use static generation to server-render a loading state, followed by fetching user data client-side.
+- Fetch user data server-side to eliminate a flash of unauthenticated content.
+
+### Static Generation
+
+Next.js automatically determines that a page is static if there are no blocking data requirements. This means the absence of `getServerSideProps` and `getInitialProps` in the page. Instead, your page can render a loading state from the server, followed by fetching the user client-side.
+
+One advantage of this pattern is it allows pages to be served from a global CDN and [preloaded](https://nextjs.org/docs/api-reference/next/link) using `<Link />`. Depending on the size of your application, [React hydration](https://reactjs.org/docs/react-dom.html#hydrate) could take some time. If you serve a loading shell from the server, this allows React to hydrate while you're fetching user data. In practice, this results in a faster TTI ([Time to Interactive](https://web.dev/interactive/)).
+
+Let's look at an example for a profile page. This will initially render a loading skeleton. Once the request for a user has finished, it will show the user's name.
+
+```jsx
+// pages/profile.js
+
+export default Profile = () => {
+  // Fetch the user client-side
+  const { user } = useUser({ redirectTo: '/login' });
+
+  // Server-render loading skeleton
+  if (!user) {
+    return <LoadingSkeleton />;
+  }
+
+  // Once the user request finishes, show the user's name
+  return <p>{user.name}</p>;
+};
+```
+
+### Server-side Rendering
+
+If you export an `async` function called `getServerSideProps` from a page, Next.js will pre-render this page on each request using the data returned by `getServerSideProps`.
+
+```jsx
+export async function getServerSideProps(context) {
+  return {
+    props: {} // Will be passed to the page component as props
+  };
+}
+```
+
+Let's transform the profile example from earlier to use server-side rendering. If there's a session, return `session` as a prop to the `Profile` component in the page. Otherwise, redirect to `/login` on the server-side. Notice there is not a loading skeleton in this example.
+
+```jsx
+// pages/profile.js
+
+export default Profile = ({ session }) => {
+  return <p>{session.user.name}</p>;
+};
+
+export async function getServerSideProps({ req, res }) {
+  // Get the user's session based on the request
+  const session = await getSession(req);
+
+  if (!session) {
+    // If no user, redirect to login
+    res.writeHead(307, { Location: '/login' });
+    res.end();
+    return { props: {} };
+  }
```

Comment:
This has some issues for client side navigations. Let's not include how to do redirects for now. Next.js will eventually support redirects here.

Reply:
What should it show in the sad path instead? Return empty props and redirect on the client-side? Login button?

```jsx
// pages/profile.js

export default Profile = ({ session }) => {
  if (!session?.user) {
    return <button>Login</button>;
  }

  return <p>{session.user.name}</p>;
};

export async function getServerSideProps({ req, res }) {
  // Get the user's session based on the request
  const session = await getSession(req);

  return { props: { session } };
}
```

Reply:
I think we should not include SSR redirects for now, doing a client side redirect in a SSR route will annoy people, but we can't also include a redirect that may have unexpected results. We can introduce this once the built-in support for redirects is out

Reply:
Updated to match the `with-iron-session` example. If still has a client-side redirect. Should we just removed that from the code snippet, or choose a different example?